### DISCUSSION TITLE
TS-XW

### DISCUSSION THREAD

**Discussion by Kutys74** [azzajess/USB-Shifter-to-Thrustmaster-Wheelbase#9] [Q&A]
Hello, I trying to work on my wheelbase, TS-XW. I am doing hardwire metodet, flashed Arduino, done wirings. When i connect to the wheelbase, arduino light is on, wheel disconnect from xbox, start rotating and again connect, but when i push buttons nothing happens in game. Can you help edit code for TS-XW?

**Comment by azzajess** [comment]
Hmmm I havent tested a ts-xw but lets assume its the same as other wheel bases for the moment. 
The fact that it resets when connected is a bit concerning. I have connected and disconnected the shifter various times while the game was running and didnt experience a wheel reset..
This leads me to believe plugging it in locks it up? It may be possible there is a short? If possible check your voltages and check for shorts when connected to the wheelbase. One thing to try is when its connected to the wheel base, connected the tx and rx to pc via ttl serial programmer and see if you get the correct shifts and power via ardruino serial monitor etc. (Pretty sure that can be done lol) Let me know how that goes.
Also make sure the actual metal chassis of the PS2 port is grounded as well, just realised I haven't made that clear enough in the documents and will add that now.
It is a strange issue....

**Reply by Kutys74** [reply]
So i have tested for shorts but cant find any, so to make sure there isn't any shorts I created a new ps2 port, and wheelbase stopped resetting, so maybe there was a short but cant find it XD. The next thing i tried was to test through the serial monitor, there some interesting things start happening, when the wheelbase is connected to the xbox i get some squares with question marks, but when i connect the wheelbase to the computer i start getting normal gears. So to me it looks like there is some kind of communication problem with the connection to the Xbox.
![Serial Monitor](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/166945816/d4c38087-4517-4639-a88e-338837902058)

**Reply by Kutys74** [reply]
So another update. I try another Arduino, unfortunately don't have another mini pro, so I use nano and now getting gears in serial monitor when connect to xbox, but nothing happening in game.

**Reply by azzajess** [reply]
Hmmmm this is strange. 
The nano doesn't have a 3v3 version so did you bypass the regulator and connect 3v3 from ps2? It sounds like your trying running it in 5v connected to the wheel?

There shouldn't be any communication come from the Xbox how odd.

The only thing I can think of is maybe shifter from ps2 port is not grounded? I'm guessing you have confirmed pin config from by testing continuity between Arduino and ps2 port? 

I'm struggling to think what might be the reason?

**Reply by Kutys74** [reply]
I use 3v3 port, on Nano. I test continuity, but not 100% sure if i do it right, can you give me instruction, how would you do it?

**Reply by azzajess** [reply]
So my bet would be to check that the pins on nano match what's connected to the ps2 pins. 
Following the image below, just make sure stuff like DIN6_6, DIN6_3 and the metal portion of the PS2 port are all the same common ground of the nano. So you may have to open the PS2 port and ensure the wires are going to their correct locations?
[Shifter Diagram](https://private-user-images.githubusercontent.com/40465008/255296957-d3780173-6143-4a0a-9bc5-9cf9eb9b1583.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyNTI0NzEsIm5iZiI6MTcxMzI1MjE3MSwicGF0aCI6Ii80MDQ2NTAwOC8yNTUyOTY5NTctZDM3ODAxNzMtNjE0My00YTBhLTliYzUtOWNmOWViOWIxNTgzLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE2VDA3MjI1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkZTVhMzBhMTAzODMyYmYyYmIyYjAzMjhhYWJlNDI0ZDViNzJkMmYzZDdjYTRkMDYxN2VhNGZlZjVmMjAwNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DZz0DSQORaXslBHHgIuKvOnY9Hg5OAgtKeScZpCSbjQ)

Maybe even chuck over some images of what your done so far? Just in case something pops out?

**Reply by Kutys74** [reply]
So I tried everything, but still got nothing. Everything seems to be done correctly. Do you have any other ideas? Because I am running out of ideas. Start thinking maybe there is problem with my wheelbase. Edit: I tried re-done everything to make sure all wirings are solder correctly, try re-flashe Arduino mini pro and nothing change. So I disassemble wheelbase and test all connection to motherboard connector and everything working fine, no shorts, continuity is good, every ground pin is ok. So I start thinking about editing code, but don't know where to start.

**Reply by azzajess** [reply]
The only possible thing I can think of is that the communication is different to that particular version of wheelbase. I don't understand how it can be different... Firmware updates and stuff? 
But a th8a shouldnt change for this product?
Unfortunately I didn't source the of th8a code and that was found in this forum
https://www.isrtv.com/forums/topic/24532-gearbox-connector-on-base/

But as far as I'm aware it involves using an oscilloscope of a th8a. But I don't have a th8a to actually test it.

Maybe a firmware update on the wheel will magically work???

Sorry I honestly don't know why it wouldn't work

**Reply by Kutys74** [reply]
Thank you, for advice anyway. I gonna do some more research and hope I can find solution. I let you know, if I find out where was problem. Maybe one last think, do you running last firmware on wheelbase?

**Reply by azzajess** [reply]
hmmm, sorry for the delay in response, i musta missed this... Um honestly not 100% sure. At the time of doing the code it was the latest, v34 for t300 rs gt wheel. version are vastly different between the different wheels...

But now that im thinking about it, it might be update related.... in the ISRTV forums (and my code) its mentioned that the codes have changed to cater for some TH sim hub. Something about sending a signal every 250ms... I have to have a think about how that should be implemented (im not a pro hahaha).. But maybe thats the solution?

**Reply by azzajess** [reply]
Try this code? Its old code that was initially in a loop. Hope it works properly lol. Its not a 250ms loop, but....maybe?
```
/*
Arduino 3.3V required or 5V with external power.

Base connector / note / Arduino pin
DIN6_1 /  nothing      /
DIN6_2 /  I2C-SCL      / A5
DIN6_3 /  /Shifter ON  / GND
DIN6_4 /  I2C-SDA      / A4
DIN6_5 /  Vdd          / RAW 3.3V
DIN6_6 /  Vss          / GND
*/

#include <Wire.h>

const int gear1 = 2;
const int gear2 = 3;
const int gear3 = 4;
const int gear4 = 5;
const int gear5 = 6;
const int gear6 = 7;
const int gear7 = 8;
const int reverse = 9; //Gear 8/Reverse
const int seqswitch = 10; //Switch on shifter used to switch to sequential



//commands seen from ociliscope in ISRTV forum - 
//apparently these codes have changed for different wheel versions however
//these worked for me...
byte command[14] = {
  0x00, // Shifter mode 0 - S / 0x80 - H
  0x0C, // Unknown
  0x01, // Unknown
  0x00, // Gear in H-mode
  0x00, // Gear in S-Mode 0x04 - center, 0x05 - down, 0x06 - up
  0x80, // Unknown
  0x80, // Unknown
  0x00, // Y cordinate
  0x00, // X cordinate
  0x00, // Unknown
  0x00, // Unknown
  0x00, // Unknown
  0x00, // Unknown
  0x00  // Unknown
};

//Sequential shift
enum position {
  center = 0x04,
  down = 0x05,
  up = 0x06
};

bool active = false;

void setup() {
  //configure pins as input with pullup resistor allowing for switches to be connected
  //between pins and ground.
  //uncomment any pins that are not being used. Also dont foget to comment the commands in loop() as well
  pinMode(gear1, INPUT_PULLUP); //Gear 1
  pinMode(gear2, INPUT_PULLUP); //Gear 2
  pinMode(gear3, INPUT_PULLUP); //Gear 3
  pinMode(gear4, INPUT_PULLUP); //Gear 4
  pinMode(gear5, INPUT_PULLUP); //Gear 5
  pinMode(gear6, INPUT_PULLUP); //Gear 6
  pinMode(gear7, INPUT_PULLUP); //Gear 7
  pinMode(reverse, INPUT_PULLUP); //Reverse
  pinMode(seqswitch, INPUT_PULLUP); //Switch (If shifter had a switch)
  Wire.begin(0x03); // join i2c bus (address optional for master)
  Serial.begin(115200);
  Serial.println("START");
  digitalWrite(13, HIGH);

  //On startup sends command neutral to initialise this. It isnt nessacary but
  //saves from moving to shift and back. Just be mindful of sequential switch i guess
  setHMode(true);
  switchHGear(0);
  sendCommand(); 
}

//H mode shifting codes
void setHMode(bool isHMode) {
  if (isHMode) {
    command[0] |= 0x80;
  } else {
    command[0] &= ~0x80;
  }
}

//Switchs gear to specificed gear number and display in serial monitor
void switchHGear(byte gear) { // Gear num 0-N, 8-R
  command[3] = (0x80 >> (8-gear));
  Serial.print("Gear ");
  Serial.print(" ");
  Serial.println(gear);
}

//Switchs gear to specificed gear number and display in serial monitor
void switchSGear(position currpos) {
  command[4] = currpos;
}

//sends command over PS2 port to T300 wheelbase
void sendCommand() {
  Wire.beginTransmission(0x01);
  Wire.write(command, 14);
  Wire.endTransmission();
}

void tryByte(byte nbyte, byte nbit) {
  command[nbyte] &= ~(0x01 << nbit-1);
  command[nbyte] |= (0x01 << nbit);
  // command[nbyte] |= 0x40;
  Serial.print("Check byte ");
  Serial.print(nbyte);
  Serial.print(" bit ");
  Serial.println(nbit);
}

void loop() {
//If statement that checks if switch is pressed to then do an action.   
while (digitalRead(seqswitch) == HIGH){

    //Gear 1
    if (digitalRead(gear1) == LOW) {
      setHMode(true);
      switchHGear(1);
      sendCommand();
    }
  
    //Gear 2
    else if (digitalRead(gear2) == LOW) {
      setHMode(true);
      switchHGear(2);
      sendCommand();
      
    }
  
    //Gear 3
    else if (digitalRead(gear3) == LOW) {
      setHMode(true);
      switchHGear(3);
      sendCommand();
      
    }
  
    //Gear 4
    else if (digitalRead(gear4) == LOW) {
      setHMode(true);
      switchHGear(4);
      sendCommand();
      
    }
  
    //Gear 5
    else if (digitalRead(gear5) == LOW) {
      setHMode(true);
      switchHGear(5);
      sendCommand();
      
    }
  
    //Gear 6
    else if (digitalRead(gear6) == LOW) {
      setHMode(true);
      switchHGear(6);
      sendCommand();
    }
  
    //Gear 7 (If your shifter does not have a 7th gear then comment this (And the digital read 8 up top))
    else if (digitalRead(gear7) == LOW){
      setHMode(true);
      switchHGear(7);
      sendCommand();
      
    }
    
    //Gear 8/Reverse
    else if (digitalRead(reverse) == LOW) {
      setHMode(true);
      switchHGear(8);
      sendCommand();
      
    }
  
    //Neutral/Gear0
    //if no other buttons are being activated then it must(?) be in neutral
    else {
      setHMode(true);
      switchHGear(0);
      sendCommand();
      active = false;
    }
}
while (digitalRead(seqswitch) == LOW){
//  //You can invert the switchSGear(xxxx) if you want to reverse sequential shifting eg up -> down/down -> up
//  //Gear 3 (Sequential shift down)
  if (digitalRead(gear3) == LOW) {
    setHMode(false);
    switchSGear(down);
    sendCommand();
    
  }

  //Gear 4 (Sequential shift up)
  else if (digitalRead(gear4) == LOW) {
    setHMode(false);
    switchSGear(up);
    sendCommand();
    
  }
       
  //Neutral/Gear0
  //if no other buttons are being activated then it must(?) be in neutral
    else  {
      setHMode(false);
      switchSGear(center);
      sendCommand();
      active = false;
    }
    
}
}
